### PR TITLE
Refactor projection and prediction heads

### LINKED ISSUE #378: Refactor Prediction/Projection Heads
# Refactor Prediction/Projection Heads

Depends on #377.

As a first step to refactoring the lightly models, we should refactor the implementation of the different prediction and projection heads. The refactored implementation should anticipate new models with currently unknown implementations of prediction/projection heads.

### PR DESCRIPTION
**Pull request by philippmwirth**
# Refactor projection and prediction heads

Closes #378.

The implementation of the prediction and projection head was abstracted in a single class `ProjectionHead`. Model specific heads are implemented by deriving from `ProjectionHead` and providing default parameters.

### CODE REVIEW

File: lightly/models/modules/heads.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class ProjectionHead(nn.Module):
+    """TODO
+    
+    """
+
+    def __init__(self, layers: List[Tuple[int, int, nn.Module, nn.Module]]):
+
+        self.ffnn = []
```
Comment (IgorSusmelj) [inline lightly/models/modules/heads.py:17]: would give it a more meaningful variable name. E.g. `self.block`, `self.head`, `self.linear`?

**Comment by IgorSusmelj** [comment]
Could you please print how a model looks like when adding the head?

create a resnet, add the head and then run `print(model)`

I'd like to see whether there is some weird output because of the layering.

Would it look like this?
```
(projectionHead)
    (linear)
    (bn)
    (relu)
    (linear)
```

File: lightly/models/modules/heads.py (hunk 1 above)
Comment (MalteEbner) [reply lightly/models/modules/heads.py:17]: I agree, I don't know what ffnn stands for. Why not `self.layers`?

File: lightly/models/modules/heads.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import torch.nn as nn
+from torch.nn.modules.batchnorm import BatchNorm1d
+
+
+class ProjectionHead(nn.Module):
+    """TODO
+    
+    """
+
+    def __init__(self, layers: List[Tuple[int, int, nn.Module, nn.Module]]):
```
Comment (MalteEbner) [inline lightly/models/modules/heads.py:15]: Is each list element a layer or other a block of (linear layer, batch norm layer, non-linear layer). If yes, than I would call it blocks.

**Review by MalteEbner — COMMENTED** [review]
I like the structure a lot! It is really great to see much more easily what makes up the difference between the different heads.

File: lightly/models/modules/heads.py (hunk 1 above)
Comment (philippmwirth) [reply lightly/models/modules/heads.py:17]: feed forward neural network... but I'll change it 😄

File: lightly/models/modules/heads.py (hunk 2 above)
Comment (philippmwirth) [reply lightly/models/modules/heads.py:15]: It's what you would call a block so I'll rename it.

**Comment by philippmwirth** [comment]
> 
> 
> Could you please print how a model looks like when adding the head?
> 
> create a resnet, add the head and then run `print(model)`
> 
> I'd like to see whether there is some weird output because of the layering.
> 
> Would it look like this?
> 
> ```
> (projectionHead)
>     (linear)
>     (bn)
>     (relu)
>     (linear)
> ```

This is the projection head without ResNet:

```
MoCoProjectionHead(
  (ffnn): Sequential(
    (0): Linear(in_features=512, out_features=512, bias=True)
    (1): ReLU()
    (2): Linear(in_features=512, out_features=256, bias=True)
  )
)
```

I don't know what you mean by "Add the head" because we've never done this. Typically we had a wrapper class (like `MoCoModel`) which would have an attribute `backbone` (the ResNet) and an attribute `projection_head`  which would be the projection head.
